src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs

```diff
@@ -704,16 +707,36 @@ public void RegisterMessageHandler(Func<Message, CancellationToken, Task> handle
                 throw new ServiceBusException(true, Resources.AmqpFieldSessionId);
             }
 
+            receivingAmqpLink.Closed += this.OnSessionReceiverLinkClosed;
             this.SessionIdInternal = tempSessionId;
             long lockedUntilUtcTicks;
             this.LockedUntilUtcInternal = receivingAmqpLink.Settings.Properties.TryGetValue(AmqpClientConstants.LockedUntilUtc, out lockedUntilUtcTicks) ? new DateTime(lockedUntilUtcTicks, DateTimeKind.Utc) : DateTime.MinValue;
         }
 
+        void OnSessionReceiverLinkClosed(object sender, EventArgs e)
```

Comment:
1) I still want the user contract to be 
User attempts operations on a Session.
If Session has expired, User Gets SessionLockLostException.
User already has code to handle SessionLockLost in other scenarios, to Close the Session and Move on.
This way we wont do anything under the covers for the user. and he suddenly discovers the Session is gone.
2) I want to continue throwing SessionLockLostException when User attempts operations on the session. Checking Closed only would mean when operations are attempted on the session, a) I would either not throw SessionLockLostException exception or b) Check that is closing and also check LinkError and throw SessionLockLostException anyway. This looked better.